share/ipld/proof.go

```diff
@@ -15,37 +15,60 @@ type Proof struct {
 // proofCollector collects proof nodes' CIDs for the construction of a shares inclusion validation
 // nmt.Proof.
 type proofCollector struct {
-	left, right []cid.Cid
+	left, right []node
+}
+
+type node struct {
+	cid   cid.Cid
+	depth int
 }
 
 func newProofCollector(maxShares int) *proofCollector {
 	// maximum possible amount of required proofs from each side is equal to tree height.
-	height := int(math.Log2(float64(maxShares)))
+	height := int(math.Log2(float64(maxShares))) + 1
 	return &proofCollector{
-		left:  make([]cid.Cid, 0, height),
-		right: make([]cid.Cid, 0, height),
+		left:  make([]node, height),
+		right: make([]node, height),
 	}
 }
 
-func (c *proofCollector) addLeft(node cid.Cid) {
-	c.left = append(c.left, node)
+func (c *proofCollector) addLeft(cid cid.Cid, depth int) {
+	c.left[depth] = node{
+		cid:   cid,
+		depth: depth,
+	}
 }
 
-func (c *proofCollector) addRight(node cid.Cid) {
-	c.right = append(c.right, node)
+func (c *proofCollector) addRight(cid cid.Cid, depth int) {
+	c.right[depth] = node{
+		cid:   cid,
+		depth: depth,
+	}
 }
 
 // Nodes returns nodes collected by proofCollector in the order that nmt.Proof validator will use
 // to traverse the tree.
 func (c *proofCollector) Nodes() []cid.Cid {
-	nodes := make([]cid.Cid, 0, len(c.left)+len(c.left))
+	cids := make([]cid.Cid, 0, len(c.left)+len(c.right))
```

Comment:
oof how did we miss this

Reply:
left and right are always the same size, check constructor